src/main_helpers.js

```diff
@@ -87,3 +89,58 @@ export function createPythonFlaskProcess(investExe) {
     logger.error('no existing invest installations found');
   }
 }
+
+export function extractZipInplace(zipFilePath) {
+  return new Promise((resolve, reject) => {
+    const extractToDir = path.dirname(zipFilePath);
+    yauzl.open(zipFilePath, { lazyEntries: true }, (err, zipfile) => {
+      if (err) throw err;
+      zipfile.readEntry();
+      zipfile.on('entry', (entry) => {
+        if (/\/$/.test(entry.fileName)) {
+          // if entry is a directory
+          fs.mkdir(path.join(extractToDir, entry.fileName), (err) => {
+            if (err) {
+              if (err.code === 'EEXIST') { } else logger.error(err);
+            }
+            zipfile.readEntry();
+          });
+        } else {
+          console.log(entry.fileName);
+          zipfile.openReadStream(entry, (err, readStream) => {
+            if (err) throw err;
+            readStream.on('end', () => {
+              zipfile.readEntry();
+            });
+            const writable = fs.createWriteStream(path.join(
+              extractToDir, entry.fileName
+            ));
+            readStream.pipe(writable);
+          });
+        }
+      });
+      zipfile.on('close', () => {
+        console.log('CLOSED')
+        resolve(true);
+      });
+    });
+  });
+}
+
+/** Check if the user has run this application before.
+ *
+ * @returns {boolean}
+ */
+export function checkFirstRun() {
+  const userDataPath = app.getPath('userData');
+  const hasRunTokenPath = path.join(userDataPath, 'app-has-run-token');
+  try {
+    if (fs.existsSync(hasRunTokenPath)) {
+      return false;
+    }
+    fs.writeFileSync(hasRunTokenPath, '');
```

Comment:
So we're just using a blank file as the indicator of whether a user has run the workbench. This seems fine, but I'm curious if there's a reason we wouldn't use an established settings tables like invest settings? Maybe the way we load and use our other settings, this is easier and less obstructive?

Reply:
I actually did exactly that - `investSettings` - for like 90% of the development of this feature, then changed it and things got much easier! The main thing is to manage this from `main` rather than `renderer`. This way `App` can simply be told via props whether it's the first run or not, it doesn't have to do any logic during rendering to figure that out. And it makes it way easier to setup front-end tests. No need to do setup with the mocked `localForage` in order to control this. Just render <App> with/without the `isFirstRun` prop.

Reply:
Sounds good to me! Thanks for walking through that decision.